controller/api/destination/endpoint_translator.go

```diff
@@ -79,33 +110,120 @@ func newEndpointTranslator(
 		availableEndpoints,
 		filteredSnapshot,
 		stream,
+		endStream,
 		log,
-		sync.Mutex{},
+		updatesQueueOverflowCounter.With(prometheus.Labels{"service": service}),
+		make(chan interface{}, updateQueueCapacity),
+		make(chan struct{}),
 	}
 }
 
 func (et *endpointTranslator) Add(set watcher.AddressSet) {
-	et.mu.Lock()
-	defer et.mu.Unlock()
+	et.enqueueUpdate(&addUpdate{set})
+}
+
+func (et *endpointTranslator) Remove(set watcher.AddressSet) {
+	et.enqueueUpdate(&removeUpdate{set})
+}
+
+func (et *endpointTranslator) NoEndpoints(exists bool) {
+	et.enqueueUpdate(&noEndpointsUpdate{exists})
+}
+
+// Add, Remove, and NoEndpoints are called from a client-go informer callback
+// and therefore must not block. For each of these, we enqueue an update in
+// a channel so that it can be processed asyncronously. To ensure that enqueuing
+// does not block, we first check to see if there is capacity in the buffered
+// channel. If there is not, we drop the update and signal to the stream that
+// it has fallen too far behind and should be closed.
+func (et *endpointTranslator) enqueueUpdate(update interface{}) {
+	select {
+	case et.updates <- update:
+		// Update has been successfully enqueued.
+	default:
+		// We are unable to enqueue because the channel does not have capacity.
+		// The stream has fallen too far behind and should be closed.
+		et.overflowCounter.Inc()
```

Comment:
Could we increment this twice during the same update? For example, if we process an EndpointSlice, and enqueue two updates:
* An Add
* A Remove.

If the Add fails because the queue is full, we'll increment the counter. Then we will increment it again when we enqueue the Remove?

Should this be incremented only once per client when it overflows? Maybe my understanding is wrong, but since we use a `service` label then all endpoint translators for a given service:port will use this counter.